opensearch/sycamore-opensearch.sh

```diff
@@ -423,15 +390,138 @@ deploy_model() {
+        info "Waiting for ${name} to deploy... ${i}/${max_reps}"
+        sleep 1
+        return 1
+    # Case 2: COMPLETED. Task is completed, so exit
+    elif [[ "${status}" == 'COMPLETED' ]]; then
+        info "Deployed ${name} successfully"
+        MODEL_ID="${model_id}"
+        return 0
+    # Case 3: FAILED. Handle some error cases, fail in unrecognized ones.
+    elif [[ "${status}" == 'FAILED' ]]; then
+        handle_deploy_error "${model_id}" "${name}" && return 0
+        deploy_task_id="$(jq -r '.task_id' "${deploy_model_log_file}")"
+    else
+        die "Unrecognized status: ${status}. Failing"
+    fi
+    return 1
+}
+
+handle_deploy_error() {
+    local model_id="$1"
+    local name="$2"
+    local deploy_model_log_file="${ARYN_STATUSDIR}/curl.better_deploy_model_task.${name}"
+    local deploy_status_file="${ARYN_STATUSDIR}/curl.deploy_task_status.${name}"
+    # handle error
+    debug "Deploy task failed for ${name}"
+    local error="$(jq -r '.error' "${deploy_status_file}")"
+    debug "${error}"
+    local worker_node="$(jq -r '.worker_node[0]' "${deploy_status_file}")"
+    local error_message="$(jq -r ".\"${worker_node}\"" <<< ${error})"
+    debug "${error_message}"
+    info "Deploy task failed for ${name}: ${error_message}"
+    # Memory Circuit Breaker error: wait 20 seconds and the try to deploy again
+    if [[ "${error_message}" = *Memory*Circuit*Breaker* ]]; then
+        local wait_time=5
+        info "Waiting for ${wait_time}s while GC runs before retrying deploy"
+        sleep ${wait_time}
+        spawn_deploy_model_task "${name}" "${model_id}"
+    # Duplicate Deploy Task error: set task id I'm watching to the RUNNING task
+    # - if no running task, then either the model is deployed or we try again
+    elif [[ "${error_message}" = "Duplicate deploy model task" ]]; then
+        _curl_json -XPOST "${BASE_URL}/_plugins/_ml/tasks/_search" \
+            -o "${deploy_task_search_file}" \
+            --data @- <<END || die "Error searching for running deploy task"
+{
+    "query": {
+        "bool": {
+            "must": [
+                {"term": {"model_id": "${model_id}"}},
+                {
+                    "bool": {
+                        "should": [
+                            {"term": {"state": "RUNNING"}},
+                            {"match_phrase": {"error": "Memory Circuit Breaker"))
+                        ]
+                    }
+                }
+            ]
+        }
+    },
+    "sort": [{"create_time": {"order": "DESC"}}]
+}
+END
+        if [[ $(jq -r '.hits.total.value' "${deploy_task_search_file}") ]]; then
+            debug "No running deploy task for ${model_id}. => check whether it's deployed"
+            model_is_deployed "${model_id}" && return 0
+            info "${model_id} failed to deploy. Try again after 1 sec"
+            spawn_deploy_model_task "${name}" "${model_id}"
+            sleep 1
```

Comment:
It looks like we generally sleep when we return 1 and the caller generally sleeps.  I know there are cases where we don't need to sleep, but they seem rare.  It may be more robust and cleaner to just have the only sleep up at L416 in the for loop.